packages/git/src/browser/dirty-diff/dirty-diff-manager.ts

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { inject, injectable, postConstruct } from 'inversify';
+import { EditorManager, EditorWidget, TextEditor } from '@theia/editor/lib/browser';
+import { Workspace } from '@theia/languages/lib/common';
+import URI from '@theia/core/lib/common/uri';
+import { DirtyDiffComputer, DirtyDiff } from './diff-computer';
+import { Emitter, Event, Disposable } from '@theia/core';
+import { DirtyDiffPreferences, DIRTYDIFF, DirtyDiffConfiguration } from './dirty-diff-preferences';
+import { PreferenceChangeEvent } from '@theia/core/lib/browser';
+import { GitResourceResolver, GIT_RESOURCE_SCHEME } from '../git-resource';
+import { WorkingDirectoryStatus, GitFileStatus, GitFileChange } from '../../common';
+import { GitRepositoryTracker } from '../git-repository-tracker';
+
+@injectable()
+export class DirtyDiffManager {
+
+    protected readonly models = new Map<string, DirtyDiffModel>();
+
+    protected readonly onDityDiffUpdateEmitter = new Emitter<DityDiffUpdate>();
+    readonly onDityDiffUpdate: Event<DityDiffUpdate> = this.onDityDiffUpdateEmitter.event;
+
+    @inject(GitRepositoryTracker) protected readonly repositoryTracker: GitRepositoryTracker;
+    @inject(GitResourceResolver) protected readonly gitResourceResolver: GitResourceResolver;
+
+    @inject(EditorManager) protected readonly editorManager: EditorManager;
+
+    @inject(Workspace) protected readonly workspace: Workspace;
+    @inject(DirtyDiffPreferences) protected readonly preferences: DirtyDiffPreferences;
+
+    @postConstruct()
+    protected async initialize() {
+        this.workspace.onDidChangeTextDocument(params => this.handleDocumentChanged(params.textDocument.uri));
+        this.preferences.onPreferenceChanged(e => this.handlePreferenceChange(e));
+        this.editorManager.onCreated(async e => await this.onEditorCreated(e));
+        this.repositoryTracker.onGitEvent(async event => await this.handleGitStatusUpdate(event.status));
+        const gitStatus = this.repositoryTracker.selectedRepositoryStatus;
+        if (gitStatus) {
+            await this.handleGitStatusUpdate(gitStatus);
+        }
+    }
+
+    protected async onEditorCreated(editorWidget: EditorWidget): Promise<void> {
+        const editor = editorWidget.editor;
```

Comment:
fyi, you can use `ReferenceCollection` to manage objects lifecycle via reference counting:
```ts
    protected readonly models = new ReferenceCollection<string, DirtyDiffModel>(
        uri => this.createNewModel(uri)
    );

    protected async onEditorCreated(editorWidget: EditorWidget): Promise<void> {
        const reference = this.models.acquire(editorWidget.editor.uri.toString());
        editorWidget.disposed(() => reference.dispose());
    }
```

Reply:
@akosyakov the `ReferenceCollection` would indeed remove those ugly details. There seems to be no way to iterate over keys of a reference collection. Would you consider this as useful for `ReferenceCollection` itself? Otherwise I'd need to track uris alongside.

Reply:
yes, it would be useful

Reply:
done.